script.module.tubed.api/resources/lib/src/httpd/handler.py

```diff
@@ -0,0 +1,201 @@
+            self.send_error(403)
+            return
+
+        if self.path.endswith('.mpd'):
+            dash_manifest = self._get_path()
+
+            LOG.debug('A manifest existence check was requested @ %s' % self.path)
+
+            if not os.path.isfile(dash_manifest):
+                LOG.error('A manifest does not exist @ %s' % self.path)
+                self.send_error(404)
+                return
+
+            self.send_response(200)
+            self.send_header('Content-Type', 'application/xml+dash')
+            self.send_header('Content-Length', str(os.path.getsize(dash_manifest)))
+            self.end_headers()
+
+            LOG.error('A manifest exists @ %s' % self.path)
+            return
+
+        LOG.error('Rejected HEAD connection to an unknown resource @ %s' % self.path)
+        self.send_error(501)
+
+    def do_POST(self):  # pylint: disable=invalid-name
+        if not self.connection_allowed():
+            LOG.warning('Rejected POST connection from %s' %
+                        ip_address(self.client_address[0]))
+            self.send_error(403)
+            return
+
+        if self.path.startswith('/widevine'):
+            LOG.debug('A widevine license request was made')
+
+            url = xbmcgui.Window(10000).getProperty('tubed-api-license_url')
+            token = xbmcgui.Window(10000).getProperty('tubed-api-license_token')
+
+            if not url:
+                LOG.error('No license url was found')
+                self.send_error(404)
+                return
+
+            if not token:
+                LOG.error('No license token was found')
+                self.send_error(403)
+                return
+
+            size_limit = None
+
+            length = int(self.headers['Content-Length'])
+            data = self.rfile.read(length)
+
+            headers = {
+                'Content-Type': 'application/x-www-form-urlencoded',
+                'Authorization': 'Bearer %s' % token
+            }
+
+            response = requests.post(url=url, headers=headers, data=data, stream=True)
+
+            content = response.raw.read(int(response.headers.get('content-length')))
+
+            header, body = content.split('\r\n\r\n'.encode('utf-8'))
+            header = header.decode('utf-8')
+
+            authorized_formats = re.search(r'^Authorized-Format-Types:\s*(?P<types>.+?)\r*$',
+                                           header, re.MULTILINE)
+            if authorized_formats:
+                authorized_types = authorized_formats.group('types').split(',')
+
+                formats = {
+                    'SD': (1280 * 528) - 1,
+                    'HD720': 1280 * 720,
+                    'HD': 7680 * 4320
+                }
+
+                if 'HD' in authorized_types:
+                    size_limit = formats['HD']
+
+                elif 'HD720' in authorized_types:
+                    if xbmc.getCondVisibility('system.platform.android') == 1:
```

Comment:
We need to think about exposing this to the python api

Reply:
which part? This is YouTube specific widevine stuff